hail/python/hail/ir/ir.py

```diff
@@ -2972,7 +2973,10 @@ def register_seeded_function(name, param_types, ret_type):
     _register(_seeded_function_registry, name, (param_types, ret_type))
 
 
-def udf(*param_types):
+T = TypeVar('T')
+
+
+def udf(*param_types) -> Callable[[Callable[..., T]], Callable[..., T]]:
```

Comment:
On `_make_dec` you added annotations on `wrapper`, but not here. Are they needed?

Reply:
Yeah, OK, so, mypy lets things slide if it finds `Any` and decorator is being a bad actor here:
```
def decorator(
    caller: Callable[..., Any], _func: Callable[..., Any] | None = ...
) -> Callable[[Callable[..., Any]], Callable[..., Any]]: ...
```

I added hail_decorator.py which applies an actually useful type to `@decorator`. I then used it here. In the end, I'm not sure how much it buys us, but all this nonsense does mean that mypy knows that `_is_snp_transition` returns `BooleanExpression`.